src/backend/distributed/executor/executor.c

```diff
@@ -1346,6 +1407,40 @@ ConnectionStateMachine(WorkerSession *session)
 }
 
 
+/*
+ * Use2PCIfParticipatingNodesExpanded ensures that the execution uses
+ * 2PC if the transaction is expanded to access a new worker node.
+ */
+static void
+Use2PCIfParticipatingNodesExpanded(WorkerSession *session)
```

Comment:
We're doing a bit of important magic in this function. I think it would help readability/naming if we split this function up and replace the call to `Use2PCIfParticipatingNodesExpanded` with something very explicit:

```c
if (IsModifyDistributedPlan(execution->distributedPlan) && TransactionModifiedDistributedTable() && 
    !ConnectionModifiedDistributedTable(session->connection))
{
    /* 
     * We already did a modification, but not on the connection that we just opened, which means we're
     * now going to make modifications over multiple connections. Activate 2PC! 
     */
    CoordinatedTransactionUse2PC();
}
```